### ISSUE TITLE
Inherited generic type inference failure

### ISSUE THREAD

**Issue by povilasb** [microsoft/pyright#1804]
**Describe the bug**

`pyright` will not check types in particular case when generics are used with inheritance.
See the example code.

**To Reproduce**

Run `pyright` on given example code.

**Expected behavior**

`mypy` says 
```
main.py:16: error: Incompatible types in assignment (expression has type "Set[str]", base class "Item" defined the type as "Set[int]")
```
That's what I would expect from `pyright` too.

**Code**

```python
from typing import Generic, TypeVar, Set, ClassVar
from pydantic.generics import GenericModel

_T = TypeVar("_T")


class Item(GenericModel, Generic[_T]):
    tags: ClassVar[Set[_T]]


str_tags = set(["a", "b", "c"])
int_tags = set([1, 2, 3])


class IntItem(Item[int]):
    tags = str_tags  # I'd expect this to be a typing error, but pyright is silent.

```

**command-line**

pyright 1.1.136
mypy 0.812

**Comment by erictraut** [comment]
Pyright does report an error if you enable the `reportIncompatibleVariableOverride` diagnostic rule. It's off by default. You can enable it globally for your project or for specific files.

**Comment by povilasb** [comment]
You're right, thank you.
Now I feel bad bothering you :)

**Comment by erictraut** [comment]
No worries, I'm happy to help.

I'm going to leave this open and investigate it further. Your example exposes an inconsistency in the way pyright is currently handling instance and class variables declared directly within the class scope and instance variables declared within methods.

You can see this in the example below.

```python
from typing import ClassVar

class A:
    cv1: ClassVar[int] = 0
    iv1: int = 0

    def __init__(self):
        self.iv2: int = 0
        self.iv3: int = 0

class B(A):
    cv1 = ""  # "reportIncompatibleVariableOverride" rule violation
    iv1 = "hi"  # "reportIncompatibleVariableOverride" rule violation

    def __init__(self):
        self.iv2 = ""  # "reportGeneralTypeIssue" rule violation
        self.iv3: str = ""  # No error
```

Perhaps "reportIncompatibleVariableOverride" should be used only in cases where a variable's type is explicitly declared (using an annotation) in both the parent and child class, but "reportGeneralTypeIssue" should be used in cases where a variable's type is declared only in the parent class.

**Comment by erictraut** [comment]
I've reviewed the various cases and updated the reportIncompatibleVariableOverride diagnostic check to handle instance variables as well as class variables. This catches the case in the example above with the "No Error" comment.

**Comment by erictraut** [comment]
This is included in pyright 1.1.143, which I just published. It will also be included in the next release of pylance.
